File: Math/java/leetcode29_Divide_Two_Integers.java
//
//Given two integers dividend and divisor, divide two integers without using multiplication, division and mod operator.
//
//Return the quotient after dividing dividend by divisor.
//
//The integer division should truncate toward zero.
//
//Example 1:
//
//Input: dividend = 10, divisor = 3
//Output: 3
//Example 2:
//
//Input: dividend = 7, divisor = -3
//Output: -2

package wenxinjie;


// be careful about data type
public class leetcode29_Divide_Two_Integers {
	public int divide(int dividend, int divisor) {
        int flag = -1;
        if ((dividend > 0 && divisor > 0) || (dividend < 0 && divisor < 0)){
            flag = 1;
        }
        long dend= Math.abs((long)dividend), visor = Math.abs((long)divisor);
        long q = helper(dend, visor);
        if (q > Integer.MAX_VALUE){
            if(flag == 1) return Integer.MAX_VALUE;
            else return Integer.MIN_VALUE;
        }
        else{
            if (flag == 1) 
                {return (int)q;}
            else 
                {return (int)(-q);}
        }        
    }
    
    public long helper(long dend, long visor){
        if (dend < visor){
            return 0;
        }
        long multi = 1, subtract = visor;
        while (subtract + subtract < dend){
            multi += multi;
            subtract += subtract;
        }
        return multi + helper(dend - subtract, visor);
    }
}


//Time: O(log(n))
//Space: O(1)
//Difficulty: medium

File: Math/python/leetcode29_Divide_Two_Integers.py

# Given two integers dividend and divisor, divide two integers without using multiplication, division and mod operator.

# Return the quotient after dividing dividend by divisor.

# The integer division should truncate toward zero.

# Example 1:

# Input: dividend = 10, divisor = 3
# Output: 3
# Example 2:

# Input: dividend = 7, divisor = -3
# Output: -2


# recursive solution, saving more time
class Solution(object):
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        if (dividend > 0 and divisor > 0)  or (dividend < 0 and divisor < 0):
            flag = 1
        else:
            flag = -1
        dend, visor = abs(dividend), abs(divisor)
        q = self.helper(dend, visor)
        if flag > 0:
            return min( q, 2** 31 - 1)
        else:
            return max( -q, -2**31)
    
    def helper(self, dend, visor):
        if dend < visor:
            return 0
        multi = 1
        subtract = visor
        while subtract + subtract < dend:
            multi = multi + multi
            subtract = subtract + subtract
        return multi + self.helper(dend - subtract, visor)


# Time: O(log(n))
# Space: O(1)
# Difficulty: medium
